server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java

```diff
@@ -19,20 +19,120 @@
 
-    public SetMailboxesResponse process(SetMailboxesRequest request) {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SetMailboxesCreationProcessor.class);
+
+    private final MailboxManager mailboxManager;
+    private final CollectionHierarchySorter<Map.Entry<MailboxCreationId, MailboxRequest>, String> collectionHierarchySorter;
+    private final MailboxUtils<Id> mailboxUtils;
+
+    @Inject
+    @VisibleForTesting
+    SetMailboxesCreationProcessor(MailboxManager mailboxManager, MailboxUtils<Id> mailboxUtils) {
+        this.mailboxManager = mailboxManager;
+        this.collectionHierarchySorter =
+            new CollectionHierarchySorter<>(
+                x -> x.getKey().getCreationId(),
+                x -> x.getValue().getParentId());
+        this.mailboxUtils = mailboxUtils;
+    }
+
+    public SetMailboxesResponse process(SetMailboxesRequest request, MailboxSession mailboxSession) {
         SetMailboxesResponse.Builder builder = SetMailboxesResponse.builder();
-        request.getCreate().entrySet().stream()
-            .forEach(entry -> builder.creation(
-                entry.getKey(),
-                Mailbox.builder().name(entry.getValue().getName()).id("serverId").build()));
+        try {
+            Map<MailboxCreationId, String> creationIdsToCreatedMailboxId = new HashMap<>();
+            collectionHierarchySorter.sortFromRootToLeaf(request.getCreate().entrySet())
+                .forEach(entry -> 
+                    createMailbox(entry.getKey(), entry.getValue(), mailboxSession, creationIdsToCreatedMailboxId, builder));
+        } catch (CycleDetectedException e) {
+            request.getCreate().entrySet()
+                .forEach(entry ->
+                    builder.notCreated(entry.getKey(),
+                            SetError.builder()
+                            .type("invalidArguments")
+                            .description("The created mailboxes introduce a cycle.")
+                            .build()));
+        }
         return builder.build();
     }
 
+    private void createMailbox(MailboxCreationId mailboxCreationId, MailboxRequest mailboxRequest, MailboxSession mailboxSession,
+            Map<MailboxCreationId, String> creationIdsToCreatedMailboxId, SetMailboxesResponse.Builder builder) {
+        try {
+            MailboxPath mailboxPath = getMailboxPath(mailboxRequest, creationIdsToCreatedMailboxId, mailboxSession);
+            mailboxManager.createMailbox(mailboxPath, mailboxSession);
+            Optional<Mailbox> mailbox = mailboxUtils.mailboxFromMailboxPath(mailboxPath, mailboxSession);
+            if (mailbox.isPresent()) {
+                builder.creation(mailboxCreationId, mailbox.get());
+                creationIdsToCreatedMailboxId.put(mailboxCreationId, mailbox.get().getId());
+            } else {
+                builder.notCreated(mailboxCreationId, SetError.builder()
```

Comment:
Why do you need to check the creation? A non throwing mailboxManager.createMailbox is not sufficient?